src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,187 @@
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(value, null, head);
+            if (head != null) {
+                head.prev = newNode;
+            } else {
+                tail = newNode;
+            }
+            head = newNode;
+            size++;
+            return;
+        }
+
+        Node<T> current;
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+
+        Node<T> newNode = new Node<>(value, current.prev, current);
+        if (current.prev != null) {
+            current.prev.next = newNode;
+        }
+        current.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds: " + index);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds: " + index);
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        T value;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds: " + index);
+        }
+        Node<T> current = head;
+        if (index == 0) {
+            value = head.value;
+            head = head.next;
+            if (head != null) {
+                head.prev = null;
+            } else {
+                tail = null;
+            }
+            size--;
+            return value;
+        }
+        if (index == size - 1) {
+            value = tail.value;
+            tail = tail.prev;
+            if (tail != null) {
+                tail.next = null;
+            } else {
+                head = null;
+            }
+            size--;
+            return value;
+        }
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        value = current.value;
+        current.prev.next = current.next;
+        current.next.prev = current.prev;
+        size--;
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
```

Comment:
The `remove(T object)` method should handle cases where the object is not found robustly, especially with null values. Ensure that the equality check logic is comprehensive.